### PR TITLE
Management of READ access to project #7945

### PR DESCRIPTION
**Pull request by vbradnitski**

### CODE REVIEW

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/AbstractProjectRolesCommand.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/UpdateProjectReadAccessCommand.java  4
- modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccess.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccessType.java  5
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectReadAccessJson.java  2
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/GetProjectReadAccessCommand.java  3
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectJson.java  2
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/AbstractProjectReadAccessCommand.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=5250142&bid=17205150)

File: modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/ProjectResource.java [hunk 1]
```diff
@@ -149,11 +157,32 @@ private CreateProjectParams createParams( final MultipartForm form )
...
+        return Optional.ofNullable( form.getAsString( PROJECT_READ_ACCESS_JSON_PROPERTY_NAME ) ).
+            map( readAccessAsString -> {
+                try
+                {
+                    return new ObjectMapper().
+                        readValue( readAccessAsString, ProjectReadAccessJson.class );
+                }
+                catch ( IOException e )
+                {
+                    LOG.error( "Cannot parse project read access.", e );
```
Comment (anatol-sialitski) [inline modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/ProjectResource.java]: @vbradnitski I guess will be better to throw `UncheckedIOException`

File: modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectName.java [hunk 2]
```diff
@@ -16,9 +16,9 @@
 
     private ProjectName( final String value )
     {
-        Preconditions.checkNotNull( value, "ProjectName cannot be null" );
-        Preconditions.checkArgument( !value.trim().isEmpty(), "ProjectName cannot be blank" );
-        Preconditions.checkArgument( value.matches( "^" + VALID_PROJECT_NAME_REGEX + "$" ), "ProjectName format incorrect: " + value );
+        Preconditions.checkNotNull( value, "Project name cannot be null" );
+        Preconditions.checkArgument( !value.trim().isEmpty(), "Project name cannot be blank" );
+        Preconditions.checkArgument( value.matches( "^" + VALID_PROJECT_NAME_REGEX + "$" ), "Project name format incorrect: " + value );
```
Comment (anatol-sialitski) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectName.java]: @vbradnitski will be better to use `Pattern.compile` and then `matches`

**Comment by jsi** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectJson.java  2
- modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccess.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/AbstractProjectRolesCommand.java  1
- modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccessType.java  5
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/UpdateProjectReadAccessCommand.java  4
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/AbstractProjectReadAccessCommand.java  1
- modules/core/core-content/src/main/java/com/enonic/xp/core/impl/project/GetProjectReadAccessCommand.java  3
- modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/json/ProjectReadAccessJson.java  2
         
```


See the complete overview on [Codacy](https://app.codacy.com/manual/enonic/xp/pullRequest?prid=5250142&bid=17205150)

File: modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/ProjectResource.java [hunk 3]
```diff
@@ -149,11 +158,31 @@ private CreateProjectParams createParams( final MultipartForm form )
...
 
+    private ProjectReadAccess getReadAccessFromForm( final MultipartForm form )
+    {
+        final ProjectReadAccess.Builder readAccess = ProjectReadAccess.create();
+
+        return Optional.ofNullable( form.getAsString( PROJECT_READ_ACCESS_JSON_PROPERTY_NAME ) ).
+            map( readAccessAsString -> {
+                try
+                {
+                    return new ObjectMapper().
```
Comment (rymsha) [inline modules/admin/admin-impl/src/main/java/com/enonic/xp/admin/impl/rest/resource/project/ProjectResource.java]: Use a shared instance of ObjectMapper

File: modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectName.java [hunk 4]
```diff
@@ -10,15 +11,15 @@
 @Beta
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectName.java]: Public API

File: modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccess.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+
+import java.util.Collection;
+import java.util.List;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableList;
+
+import com.enonic.xp.security.PrincipalKey;
+
+public final class ProjectReadAccess
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccess.java]: Shall it really be a part of API?

File: modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccessType.java [hunk 6]
```diff
@@ -0,0 +1,24 @@
+package com.enonic.xp.project;
+
+public enum ProjectReadAccessType
```
Comment (rymsha) [inline modules/core/core-api/src/main/java/com/enonic/xp/project/ProjectReadAccessType.java]: Shall it really be a part of API?

**Pull request closed without merging**